src/ocaml/typing/printtyp.ml

```diff
@@ -15,2664 +15,66 @@
-        (Errortrace.Subtype.{trace = tr_sub; unification_trace = tr_unif})
-        txt1 =
-    wrap_printing_env ~error:true env (fun () ->
-      reset ();
-      let tr_sub = prepare_trace prepare_expansion tr_sub in
-      let tr_unif = prepare_unification_trace prepare_expansion tr_unif in
-      let keep_first = match tr_unif with
-        | [Obj _ | Variant _ | Escape _ ] | [] -> true
-        | _ -> false in
-      fprintf ppf "@[<v>%a"
-        (trace filter_subtype_trace subtype_get_diff true keep_first txt1)
-        tr_sub;
-      if tr_unif = [] then fprintf ppf "@]" else
-        let mis = mismatch (dprintf "Within this type") env tr_unif in
-        fprintf ppf "%a%t%t@]"
-          (trace filter_trace unification_get_diff false
-             (mis = None) "is not compatible with type") tr_unif
-          (explain mis)
-          Conflicts.print_explanations
-    )
-end
-
-let report_ambiguous_type_error ppf env tp0 tpl txt1 txt2 txt3 =
-  wrap_printing_env ~error:true env (fun () ->
-    reset ();
-    let tp0 = trees_of_type_path_expansion tp0 in
-      match tpl with
-      [] -> assert false
-    | [tp] ->
-        fprintf ppf
-          "@[%t@;<1 2>%a@ \
-             %t@;<1 2>%a\
-           @]"
-          txt1 type_path_expansion (trees_of_type_path_expansion tp)
-          txt3 type_path_expansion tp0
-    | _ ->
-        fprintf ppf
-          "@[%t@;<1 2>@[<hv>%a@]\
-             @ %t@;<1 2>%a\
-           @]"
-          txt2 type_path_list (List.map trees_of_type_path_expansion tpl)
-          txt3 type_path_expansion tp0)
-
-(* Adapt functions to exposed interface *)
-let tree_of_path = tree_of_path None
-let tree_of_modtype = tree_of_modtype ~ellipsis:false
-let type_expansion mode ppf ty_exp =
-  type_expansion ppf (trees_of_type_expansion mode ty_exp)
-let tree_of_type_declaration ident td rs =
-  with_hidden_items [{hide=true; ident}]
-    (fun () -> tree_of_type_declaration ident td rs)
-
-let shorten_type_path env p =
-  wrap_printing_env env
-    (fun () -> best_type_path_simple p)
-
-let shorten_module_type_path env p =
-  wrap_printing_env env
-    (fun () -> best_module_type_path p)
+    Doc.Naming_context.with_arg id
+      (fun () -> functor_parameters ~sep custom_printer q)
 
-let shorten_module_path env p =
-  wrap_printing_env env
-    (fun () -> best_module_path p)
+let modtype_declaration = Fmt.compat1 Doc.modtype_declaration
+let class_type = Fmt.compat Doc.class_type
+let class_declaration = Fmt.compat1 Doc.class_declaration
+let cltype_declaration = Fmt.compat1 Doc.cltype_declaration
+let type_expansion = Fmt.compat1 Doc.type_expansion
+let printed_signature = Fmt.compat1 Doc.printed_signature
 
-let shorten_class_type_path env p =
-  wrap_printing_env env
-    (fun () -> best_class_type_path_simple p)
 
-let () =
-  Env.shorten_module_path := shorten_module_path
+let () = Env.print_longident := Doc.longident
+let () = Env.print_path := Doc.path
+let () = Btype.print_raw := Doc.raw_type_expr
```

Comment:
This part looks weird: the debugging printers now lives in Rawprinttyp which avoids the cyclic dependency on Btype. Normally, it should be doable to remove all raw functions in Printtyp and Btype.

Reply:
@xvw could you explain again what was your strategy with merging the printing changes ?

Reply:
@Octachron the print_raw forward declaration in btype was actually unused, and removed in upstream so I reflected that change in [13407ca](https://github.com/ocaml/merlin/pull/1850/commits/13407cad0047e4458c7efc5681877cffa05dc663). 

Was your comment also referring to the others forward declarations in Env ?